### ISSUE TITLE
Error handling for temporary connectivity issues

### ISSUE THREAD

**Issue by tomykaira** [embulk/embulk-input-gcs#20]
Hi, I am happy with this plugin, but downloads are sometimes disconnected by Google.

Is it possible to insert exponential backoff to downloading and make embulk run stable?

Stacktrace:

```
2016-08-02 14:09:12.500 +0000 [INFO] (0001:transaction): {done:  1 / 1, running: 0}
org.embulk.exec.PartialExecutionException: java.lang.RuntimeException: javax.net.ssl.SSLException: Connection has been shutdown: javax.net.ssl.SSLException: SSL peer shut down incorrectly
    at org.embulk.exec.BulkLoader$LoaderState.buildPartialExecuteException(org/embulk/exec/BulkLoader.java:363)
    at org.embulk.exec.BulkLoader.doRun(org/embulk/exec/BulkLoader.java:572)
    at org.embulk.exec.BulkLoader.access$000(org/embulk/exec/BulkLoader.java:33)
    at org.embulk.exec.BulkLoader$1.run(org/embulk/exec/BulkLoader.java:374)
    at org.embulk.exec.BulkLoader$1.run(org/embulk/exec/BulkLoader.java:370)
    at org.embulk.spi.Exec.doWith(org/embulk/spi/Exec.java:25)
    at org.embulk.exec.BulkLoader.run(org/embulk/exec/BulkLoader.java:370)
    at org.embulk.EmbulkEmbed.run(org/embulk/EmbulkEmbed.java:180)
    at java.lang.reflect.Method.invoke(java/lang/reflect/Method.java:606)
    at RUBY.run(uri:classloader:/embulk/runner.rb:84)
    at RUBY.run(uri:classloader:/embulk/command/embulk_run.rb:306)
    at RUBY.<top>(uri:classloader:/embulk/command/embulk_main.rb:2)
    at org.jruby.RubyKernel.require(org/jruby/RubyKernel.java:937)
    at usr.bin.embulk.embulk.command.embulk_bundle.<top>(file:/usr/bin/embulk!/embulk/command/embulk_bundle.rb:30)
    at java.lang.invoke.MethodHandle.invokeWithArguments(java/lang/invoke/MethodHandle.java:599)
    at org.embulk.cli.Main.main(org/embulk/cli/Main.java:23)
Caused by: java.lang.RuntimeException: javax.net.ssl.SSLException: Connection has been shutdown: javax.net.ssl.SSLException: SSL peer shut down incorrectly
    at org.embulk.spi.util.InputStreamFileInput.poll(org/embulk/spi/util/InputStreamFileInput.java:151)
    at org.embulk.spi.util.FileInputInputStream.nextBuffer(org/embulk/spi/util/FileInputInputStream.java:90)
    at org.embulk.spi.util.FileInputInputStream.read(org/embulk/spi/util/FileInputInputStream.java:56)
    at java.util.zip.InflaterInputStream.fill(java/util/zip/InflaterInputStream.java:238)
    at java.util.zip.InflaterInputStream.read(java/util/zip/InflaterInputStream.java:158)
    at java.util.zip.GZIPInputStream.read(java/util/zip/GZIPInputStream.java:116)
    at org.embulk.spi.util.InputStreamFileInput.poll(org/embulk/spi/util/InputStreamFileInput.java:142)
    at org.embulk.spi.util.FileInputInputStream.nextBuffer(org/embulk/spi/util/FileInputInputStream.java:90)
    at org.embulk.spi.util.FileInputInputStream.read(org/embulk/spi/util/FileInputInputStream.java:56)
    at sun.nio.cs.StreamDecoder.readBytes(sun/nio/cs/StreamDecoder.java:283)
    at sun.nio.cs.StreamDecoder.implRead(sun/nio/cs/StreamDecoder.java:325)
    at sun.nio.cs.StreamDecoder.read(sun/nio/cs/StreamDecoder.java:177)
    at java.io.InputStreamReader.read(java/io/InputStreamReader.java:184)
    at java.io.BufferedReader.fill(java/io/BufferedReader.java:154)
    at java.io.BufferedReader.readLine(java/io/BufferedReader.java:317)
    at java.io.BufferedReader.readLine(java/io/BufferedReader.java:382)
    at org.embulk.spi.util.LineDecoder.poll(org/embulk/spi/util/LineDecoder.java:94)
    at org.embulk.parser.jsonl.JsonlParserPlugin.run(org/embulk/parser/jsonl/JsonlParserPlugin.java:155)
    at org.embulk.spi.FileInputRunner.run(org/embulk/spi/FileInputRunner.java:154)
    at org.embulk.spi.util.Executors.process(org/embulk/spi/util/Executors.java:67)
    at org.embulk.spi.util.Executors.process(org/embulk/spi/util/Executors.java:42)
    at org.embulk.exec.LocalExecutorPlugin$DirectExecutor$1.call(org/embulk/exec/LocalExecutorPlugin.java:184)
    at org.embulk.exec.LocalExecutorPlugin$DirectExecutor$1.call(org/embulk/exec/LocalExecutorPlugin.java:180)
    at java.util.concurrent.FutureTask.run(java/util/concurrent/FutureTask.java:262)
    at java.util.concurrent.ThreadPoolExecutor.runWorker(java/util/concurrent/ThreadPoolExecutor.java:1145)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(java/util/concurrent/ThreadPoolExecutor.java:615)
    at java.lang.Thread.run(java/lang/Thread.java:745)
Caused by: javax.net.ssl.SSLException: Connection has been shutdown: javax.net.ssl.SSLException: SSL peer shut down incorrectly
    at sun.security.ssl.SSLSocketImpl.checkEOF(SSLSocketImpl.java:1508)
    at sun.security.ssl.AppInputStream.available(AppInputStream.java:59)
    at java.io.BufferedInputStream.available(BufferedInputStream.java:399)
    at sun.net.www.MeteredStream.available(MeteredStream.java:170)
    at sun.net.www.http.KeepAliveStream.close(KeepAliveStream.java:85)
    at java.io.FilterInputStream.close(FilterInputStream.java:181)
    at sun.net.www.protocol.http.HttpURLConnection$HttpInputStream.close(HttpURLConnection.java:3137)
    at java.io.FilterInputStream.close(FilterInputStream.java:181)
    at org.embulk.spi.util.ResumableInputStream.reopen(ResumableInputStream.java:35)
    at org.embulk.spi.util.ResumableInputStream.read(ResumableInputStream.java:78)
    at org.embulk.spi.util.InputStreamFileInput.poll(InputStreamFileInput.java:142)
    at org.embulk.spi.util.FileInputInputStream.nextBuffer(FileInputInputStream.java:90)
    at org.embulk.spi.util.FileInputInputStream.read(FileInputInputStream.java:56)
    at java.util.zip.InflaterInputStream.fill(InflaterInputStream.java:238)
    at java.util.zip.InflaterInputStream.read(InflaterInputStream.java:158)
    at java.util.zip.GZIPInputStream.read(GZIPInputStream.java:116)
    at org.embulk.spi.util.InputStreamFileInput.poll(InputStreamFileInput.java:142)
    at org.embulk.spi.util.FileInputInputStream.nextBuffer(FileInputInputStream.java:90)
    at org.embulk.spi.util.FileInputInputStream.read(FileInputInputStream.java:56)
    at sun.nio.cs.StreamDecoder.readBytes(StreamDecoder.java:283)
    at sun.nio.cs.StreamDecoder.implRead(StreamDecoder.java:325)
    at sun.nio.cs.StreamDecoder.read(StreamDecoder.java:177)
    at java.io.InputStreamReader.read(InputStreamReader.java:184)
    at java.io.BufferedReader.fill(BufferedReader.java:154)
    at java.io.BufferedReader.readLine(BufferedReader.java:317)
    at java.io.BufferedReader.readLine(BufferedReader.java:382)
    at org.embulk.spi.util.LineDecoder.poll(LineDecoder.java:94)
    at org.embulk.parser.jsonl.JsonlParserPlugin.run(JsonlParserPlugin.java:155)
    at org.embulk.spi.FileInputRunner.run(FileInputRunner.java:154)
    at org.embulk.spi.util.Executors.process(Executors.java:67)
    at org.embulk.spi.util.Executors.process(Executors.java:42)
    at org.embulk.exec.LocalExecutorPlugin$DirectExecutor$1.call(LocalExecutorPlugin.java:184)
    at org.embulk.exec.LocalExecutorPlugin$DirectExecutor$1.call(LocalExecutorPlugin.java:180)
    at java.util.concurrent.FutureTask.run(FutureTask.java:262)
    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
    at java.lang.Thread.run(Thread.java:745)
Caused by: javax.net.ssl.SSLException: SSL peer shut down incorrectly
    at sun.security.ssl.InputRecord.readV3Record(InputRecord.java:557)
    at sun.security.ssl.InputRecord.read(InputRecord.java:509)
    at sun.security.ssl.SSLSocketImpl.readRecord(SSLSocketImpl.java:946)
    at sun.security.ssl.SSLSocketImpl.readDataRecord(SSLSocketImpl.java:903)
    at sun.security.ssl.AppInputStream.read(AppInputStream.java:102)
    at java.io.BufferedInputStream.read1(BufferedInputStream.java:273)
    at java.io.BufferedInputStream.read(BufferedInputStream.java:334)
    at sun.net.www.MeteredStream.read(MeteredStream.java:134)
    at java.io.FilterInputStream.read(FilterInputStream.java:133)
    at sun.net.www.protocol.http.HttpURLConnection$HttpInputStream.read(HttpURLConnection.java:3067)
    at com.google.api.client.http.javanet.NetHttpResponse$SizeValidatingInputStream.read(NetHttpResponse.java:169)
    at org.embulk.spi.util.ResumableInputStream.read(ResumableInputStream.java:74)
    ... 27 more

Error: java.lang.RuntimeException: javax.net.ssl.SSLException: Connection has been shutdown: javax.net.ssl.SSLException: SSL peer shut down incorrectly
```

**Comment by sakama** [comment]
@tomykaira Thank you for your reporting.
Are you using recent version of embulk-input-gcs?

I implemented [retry logic with exponential backoff](https://github.com/embulk/embulk-input-gcs/blob/master/src/main/java/org/embulk/input/gcs/GcsFileInputPlugin.java#L315) at [v0.1.1](https://github.com/embulk/embulk-input-gcs/blob/master/ChangeLog#L9).

**Comment by tomykaira** [comment]
Yes, my embulk gemfile contains `embulk-input-gcs (0.2.0)`.
I did not dig into the problem, but did this exception come from other than v0.1.1 approached?

**Comment by sakama** [comment]
hmm... It seems plugin failed to handle SSLException at [around here](https://github.com/embulk/embulk-input-gcs/blob/master/src/main/java/org/embulk/input/gcs/GcsFileInputPlugin.java#L315-L366).
But I have no idea which code I should fix.

**Comment by sakama** [comment]
I added retry logic #21 that I wrote for other embulk plugin before and released [v0.2.1](https://rubygems.org/gems/embulk-input-gcs/versions/0.2.1).
This fix will improve connection stability but seems to have no relationship with your error:)

**Comment by tomykaira** [comment]
Hmmm, thanks. This error is rare, but I will introduce 0.2.1 and try for some period.
